File: AoC2015/Day1/aoc2015_day1_part_b.js
/**
 * Now, given the same instructions, find the position 
 * of the first character that causes him to enter the 
 * basement (floor -1).
 */

/** Functions */
function findBasementPosition(instructions) {
    /**
     * Find position of the first character that causes 
     * santa to enter the basement floor (-1).
     */
    // Santa starts at floor zero
    var floorNumber = 0, instructionPosition = 1;
    // Iterate over instructions
    for (instruction of instructions) {
        if (instruction === '(') {
            floorNumber++;
        } else if (instruction === ')') {
            floorNumber--;
            if (floorNumber < 0) return instructionPosition;
        }
        instructionPosition++;
    }
}

/** Driver Program */
// Santa's instructions
instructions = null;
// Require the filesystem module
const fs = require('fs');
// Require the assertions module
const assert = require("assert");
// Read contents of file as a string synchronously
try {
    instructions = fs.readFileSync('aoc2015_day1_input.txt', 'utf8');
} catch (err) {
    if (err) console.error(err);
}

/** Assertions */
// Some assertions to check the function is working correctly
try {
    // Sample I/O assertions
    assert.equal(findBasementPosition(")"), 1, "TEST1");
    assert.equal(findBasementPosition("()())"), 5, "TEST2");
    // Final assertion on challenge input
    assert.equal(findBasementPosition(instructions), 1783, "TESTN");
} catch (err) {
    console.error(err);
}

File: AoC2015/Day1/aoc2015_day1_part_b.rb
# Now, given the same instructions, find the position 
# of the first character that causes him to enter the 
# basement (floor -1).

def find_basement_position(instructions)
    # Find position of the first character that causes 
    # santa to enter the basement floor (-1).
    # Santa starts at floor zero
    # Instruction position starts at one
    floor_number, instruction_position = 0, 1
    # Iterate over instructions
    instructions.chars.each do |instruction|
        if instruction == '('
            floor_number += 1
        elsif instruction == ')'
            floor_number -= 1
            if floor_number < 0
                return instruction_position
            end
        end
        instruction_position += 1
    end
end

# Driver program
instructions = IO.read("aoc2015_day1_input.txt") # Santa's instructions
# Some assertions to check the function is working correctly
puts "TEST1: #{find_basement_position(")") == 1}"
puts "TEST2: #{find_basement_position("()())") == 5}"
# Final assertion on challenge input
puts "TESTN: #{find_basement_position(instructions) == 1783}"